Add comments to document blocks

tododeni opened the pull request:
<!--
### Requirements for making a pull request

Thank you for contributing to our project!

Please fill out the template below to help the project maintainers review it as fast as possible and include your contribution to the project.
-->

### What does it fix?
- Adds comments to document blocks
- CRUD operations for comments
- Option to list all comments of a document block
- Unit tests for all Comment operations

Closes # [https://github.com/code4romania/expert-consultation-backend/issues/27](url)

<!-- Please mention the main changes this PR brings. -->

### How has it been tested?

<!-- Please describe the tests that you ran to verify your changes. -->
Unit tests for all operations
<!-- If applicable, mention any known issues with the pull request -->
For better testing, a new issue should be opened regarding a basic mock population of the db, so that tests can be performed for each call.

tododeni committed:
CRUD for Comment

tododeni committed:
CRUD for Comment (update and delete)

tododeni committed:
Tests for CRUD comment

tododeni committed:
List all comments for a block + tests

tododeni committed:
Cleanup

georgebejan commented on src/main/java/ro/code4/expertconsultation/comment/model/dto/CommentDto.java:
```diff
@@ -0,0 +1,17 @@
+@AllArgsConstructor
+@EqualsAndHashCode(callSuper = false)
+public class CommentDto extends IdentifiableDto {
+    private Timestamp creationTime;
```
Please use `LocalDateTime` instead of `Timestamp`. We don't want to depend on JDBC API for dates in our application

tododeni replied:
done

georgebejan commented on src/main/java/ro/code4/expertconsultation/comment/model/persistence/Comment.java:
```diff
@@ -0,0 +1,32 @@
+public class Comment extends Identifiable {
+
+
+    private Timestamp creationTime;
```
Please use `LocalDateTime` instead of `Timestamp`. We don't want to depend on JDBC API for dates in our application

tododeni replied:
done

georgebejan commented on src/main/java/ro/code4/expertconsultation/comment/service/impl/CommentServiceImpl.java:
```diff
@@ -0,0 +1,101 @@
+    @Transactional(readOnly = true)
+    public List<CommentDto> list(final Long documentBlockId) {
+        if (documentBlockId == null) {
+            throw new InvalidArgumentException("Arguments cannot be null");
```
please use `ExpertConsultationException` to throw exceptions (see `FilesystemStorageService` for example)

tododeni replied:
done

georgebejan commented on src/main/java/ro/code4/expertconsultation/comment/service/impl/CommentServiceImpl.java:
```diff
@@ -0,0 +1,101 @@
+    @Transactional(readOnly = true)
+    public CommentDto get(Long id) {
+        if (id == null) {
+            throw new InvalidArgumentException("Arguments cannot be null");
```
please use `ExpertConsultationException` to throw exceptions (see `FilesystemStorageService` for example)

tododeni replied:
done

georgebejan commented on src/main/java/ro/code4/expertconsultation/comment/service/impl/CommentServiceImpl.java:
```diff
@@ -0,0 +1,101 @@
+    public CommentDto create(Long userId, Long documentId, Long blockId, CommentDto commentDto) {
+
+        if (((userId == null) || (documentId == null) || (blockId == null) || (commentDto == null))) {
+            throw new InvalidArgumentException("Arguments cannot be null");
```
please use `ExpertConsultationException` to throw exceptions (see `FilesystemStorageService` for example)

tododeni replied:
done

georgebejan commented on src/main/java/ro/code4/expertconsultation/comment/service/impl/CommentServiceImpl.java:
```diff
@@ -0,0 +1,101 @@
+    @Transactional
+    public CommentDto update(Long commentId, CommentDto commentDto) {
+        if ((commentId == null) || (commentDto == null)) {
+            throw new InvalidArgumentException("Arguments cannot be null");
```
please use `ExpertConsultationException` to throw exceptions (see `FilesystemStorageService` for example)

tododeni replied:
done

georgebejan commented on src/main/java/ro/code4/expertconsultation/comment/service/impl/CommentServiceImpl.java:
```diff
@@ -0,0 +1,101 @@
+    @Transactional
+    public void delete(Long commentId) {
+        if (commentId == null) {
+            throw new InvalidArgumentException("Arguments cannot be null");
```
please use `ExpertConsultationException` to throw exceptions (see `FilesystemStorageService` for example)

tododeni replied:
done

georgebejan commented on src/main/java/ro/code4/expertconsultation/document/service/impl/DocumentBlockServiceImpl.java:
```diff
@@ -70,4 +70,9 @@ public DocumentBlockDto update(final Long documentId,
         documentBlockRepository.deleteById(documentBlockId);
     }
+
+    public DocumentBlock getEntity(final Long documentId, final Long documentBlockId) {
```
Why is `documentId` needed here?

tododeni replied:
not needed. My bad. Removed from unnecessary places

georgebejan commented on src/main/java/ro/code4/expertconsultation/exception/InvalidArgumentException.java:
```diff
@@ -0,0 +1,8 @@
+package ro.code4.expertconsultation.exception;
+
+public class InvalidArgumentException extends RuntimeException{
```
Check `ExpertConsultationException`

tododeni replied:
done

georgebejan commented on src/main/java/ro/code4/expertconsultation/document/service/impl/DocumentBlockServiceImpl.java:
```diff
@@ -70,4 +70,9 @@ public DocumentBlockDto update(final Long documentId,
     }
+
+    public DocumentBlock getEntity(final Long documentId, final Long documentBlockId) {
+        return documentBlockRepository.findById(documentBlockId)
```
Use `documentBlockRepository.getById` since we don't need to fetch the whole entity, only the reference to the entity to be set in the comments.

tododeni replied:
done

georgebejan commented on src/main/java/ro/code4/expertconsultation/user/service/impl/UserServiceImpl.java:
```diff
@@ -66,4 +66,9 @@ public UserDto update(final Long userId, final UserDto userDto) {
+
+    @Override
+    public User getEntity(final Long userId) {
+        return userRepository.findById(userId).orElseThrow(EntityNotFoundException::new);
```
Use `userRepository.getById` since we don't need to fetch the whole entity, only the reference to the entity to be set in the comments
Check https://www.javacodemonk.com/difference-between-getone-and-findbyid-in-spring-data-jpa-3a96c3ff

tododeni replied:
Thanks :). Great info. 👍  Done

georgebejan commented on src/test/java/ro/code4/expertconsultation/TestUtils.java:
```diff
@@ -0,0 +1,100 @@
+import java.sql.Timestamp;
+import java.util.List;
+
+public class TestUtils {
```
Separate these methods to different factories (`CommentFactory`, etc..)

tododeni replied:
done

tododeni committed:
Changed TimeStamp to LocalDateTime
Create Factories for testing
Replace .findById with getById
Replace (and remove) InvalidArgumentException with ExpertConsultationException
Adjusted testing to new modifications

tododeni committed:
Corrected and removed unnecessary documentId

georgebejan commented on src/main/java/ro/code4/expertconsultation/comment/controller/CommentController.java:
```diff
@@ -0,0 +1,47 @@
+        return commentService.get(id);
+    }
+
+    @GetMapping("/{blockId}")
```
you already have a root url `/api/documents/{documentId}/blocks/{blockId}/comments` ... adding `blockId` to the `GetMapping` annotation will result in the following url for the method `/api/documents/{documentId}/blocks/{blockId}/comments/{blockId}`

tododeni replied:
modified

tododeni committed:
Modify request mapping for Comments
Create CommentCreationRequest
Update Comment service
Update Comment Service Tests
Create constant for NULL_ARGUMENTS_RECEIVED
